File: solutions/python/complex-numbers/1/complex_numbers.py
from math import exp, cos, sin


class ComplexNumber:
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        if isinstance(other, (int, float)):
            return self.real == other and self.imaginary == 0
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        if isinstance(other, (int, float)):
            return ComplexNumber(self.real + other, self.imaginary)
        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __radd__(self, other):
        return self.__add__(other)

    def __mul__(self, other):
        if isinstance(other, (int, float)):
            return ComplexNumber(self.real * other, self.imaginary * other)
        return ComplexNumber(
            self.real * other.real - self.imaginary * other.imaginary,
            self.real * other.imaginary + self.imaginary * other.real
        )

    def __rmul__(self, other):
        return self.__mul__(other)

    def __sub__(self, other):
        if isinstance(other, (int, float)):
            return ComplexNumber(self.real - other, self.imaginary)
        return ComplexNumber(self.real - other.real, self.imaginary - other.imaginary)

    def __rsub__(self, other):
        return ComplexNumber(other - self.real, -self.imaginary)

    def __truediv__(self, other):
        if isinstance(other, (int, float)):
            return ComplexNumber(self.real / other, self.imaginary / other)

        denominator = other.real ** 2 + other.imaginary ** 2
        new_real = (self.real * other.real + self.imaginary * other.imaginary) / denominator
        new_imaginary = (self.imaginary * other.real - self.real * other.imaginary) / denominator
        return ComplexNumber(new_real, new_imaginary)

    def __rtruediv__(self, other):
        denominator = self.real ** 2 + self.imaginary ** 2
        return ComplexNumber(
            (other * self.real) / denominator,
            (-other * self.imaginary) / denominator
        )

    def __abs__(self):
        return (self.real ** 2 + self.imaginary ** 2) ** 0.5

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        magnitude = exp(self.real)
        angle = self.imaginary
        return ComplexNumber(magnitude * cos(angle), magnitude * sin(angle))

File: solutions/cpp/complex-numbers/1/complex_numbers.cpp
#include "complex_numbers.h"

#include <cmath>

namespace complex_numbers {
    Complex::Complex(const double x, const double y) : real_{x}, imag_{y} {
    };

    double Complex::real() const {
        return real_;
    }

    double Complex::imag() const {
        return imag_;
    }

    double Complex::abs() const {
        return std::sqrt(real_ * real_ + imag_ * imag_);
    }

    Complex Complex::conj() const {
        return {real_, -imag_};
    }

    Complex Complex::exp() const {
        const double magnitude = std::exp(real_);
        const double angle = imag_;
        return {magnitude * cos(angle), magnitude * sin(angle)};
    }

    Complex operator+(const Complex &c1, const Complex &c2) {
        return {c1.real() + c2.real(), c1.imag() + c2.imag()};
    }

    Complex operator+(const double x, const Complex &c) {
        return {x + c.real(), c.imag()};
    }

    Complex operator+(const Complex &x, double y) {
        return {x.real() + y, x.imag()};
    }

    Complex operator-(const Complex &c1, const Complex &c2) {
        return {c1.real() - c2.real(), c1.imag() - c2.imag()};
    }

    Complex operator-(const double x, const Complex &c) {
        return {x - c.real(), -c.imag()};
    }

    Complex operator-(const Complex &x, double y) {
        return {x.real() - y, x.imag()};
    }

    Complex operator*(const Complex &c1, const Complex &c2) {
        return {c1.real() * c2.real() - c1.imag() * c2.imag(), c1.real() * c2.imag() + c1.imag() * c2.real()};
    }

    Complex operator*(const double x, const Complex &c) {
        return {x * c.real(), x * c.imag()};
    }

    Complex operator*(const Complex &x, double y) {
        return {x.real() * y, x.imag() * y};
    }

    Complex operator/(const Complex &c1, const Complex &c2) {
        const double dem = c2.real() * c2.real() + c2.imag() * c2.imag();
        return {
            (c1.real() * c2.real() + c1.imag() * c2.imag()) / dem, (c1.imag() * c2.real() - c1.real() * c2.imag()) / dem
        };
    }

    Complex operator/(const double x, const Complex &c) {
        return Complex{x,0} / c;
    }

    Complex operator/(const Complex &c, double y) {
        return {c.real() / y, c.imag() / y};
    }
} // namespace complex_numbers
